sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java

```diff
@@ -202,6 +261,220 @@ private Session session() {
+        } else {
+            sourceIndexName = table.ident().indexNameOrAlias();
+            sourceIndexAlias = null;
+        }
+
+        ClusterState currentState = clusterService.state();
+
+        final Integer targetNumberOfShards = getValidNumberOfShards(analysis.tableParameter().settings());
+        validateForResizeRequest(currentState, sourceIndexName, targetNumberOfShards);
+
+        // pickup node
+        DiscoveryNode resizeNode = getNodeForResize(currentState);
+
+        List<ChainableAction<Long>> actions = new ArrayList<>(7);
+        // change allocation
+        actions.add(new ChainableAction<>(
+            () -> initAllocationToNode(resizeNode.getName(), sourceIndexName),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+
+        // ensure index allocation
+        actions.add(new ChainableAction<>(
+            () -> ensureIndexAllocation(sourceIndexName),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+
+        // shrink index
+        String targetIndexName = SHRINK_PREFIX + sourceIndexName;
+        actions.add(new ChainableAction<>(
+            () -> shrinkIndex(sourceIndexName, sourceIndexAlias, targetIndexName, targetNumberOfShards),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+
+        // index rename
+        final String backupIndexName = BACKUP_PREFIX + sourceIndexName;
+
+        List<RenameIndexAction> renameIndexActionList = new ArrayList<>(2);
+        renameIndexActionList.add(new RenameIndexAction(sourceIndexName, backupIndexName));
+        renameIndexActionList.add(new RenameIndexAction(targetIndexName, sourceIndexName));
+
+        actions.add(new ChainableAction<>(
+            () -> renameIndicesBulk(renameIndexActionList),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+
+        // delete backup index
+        actions.add(new ChainableAction<>(
+            () -> deleteIndex(backupIndexName),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+
+        // revert index allocation
+        actions.add(new ChainableAction<>(
+            () -> initRevertAllocation(sourceIndexName),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+
+        // ensure index allocation
+        actions.add(new ChainableAction<>(
+            () -> ensureIndexAllocation(sourceIndexName),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+        return ChainableActions.run(actions);
+    }
+
+    private static void validateForResizeRequest(ClusterState clusterState, String indexName, int targetNumberOfShards) {
+        final IndexMetaData indexMetaData = clusterState.metaData().index(indexName);
+        validateNumberOfShardsForResize(indexMetaData, targetNumberOfShards);
+        validateReadOnlyIndexForResize(indexMetaData);
+    }
+
+    @VisibleForTesting
+    static Integer getValidNumberOfShards(final Settings settings) {
```

Comment:
why "ValidNumberOfShards"? (to me it seems as it's just reading the setting)

Reply:
Renamed it in https://github.com/crate/crate/pull/7992/files#diff-b5f2ccdb9fb89e66959fddf9c1a0cec2R343